File: hard/0295_find_median_from_data_stream/solution.py
import heapq


class MedianFinder:
    def __init__(self):
        self.maxHeap = []
        self.minHeap = []

    def addNum(self, num: int) -> None:
        if (
            not self.minHeap
            or (self.maxHeap and num > -self.maxHeap[0])
            or (self.minHeap and num > self.minHeap[0])
        ):
            heapq.heappush(self.minHeap, num)
        else:
            heapq.heappush(self.maxHeap, -num)

        if len(self.minHeap) - len(self.maxHeap) == 2:
            heapq.heappush(self.maxHeap, -self.minHeap[0])
            heapq.heappop(self.minHeap)
        elif len(self.maxHeap) - len(self.minHeap) == 1:
            heapq.heappush(self.minHeap, -self.maxHeap[0])
            heapq.heappop(self.maxHeap)

    def findMedian(self) -> float:
        if (len(self.maxHeap) + len(self.minHeap)) % 2 == 0:
            return (-self.maxHeap[0] + self.minHeap[0]) / 2
        else:
            return self.minHeap[0]

File: hard/0295_find_median_from_data_stream/solution.cpp
#include <queue>
#include <vector>

class MedianFinder {
private:
  std::priority_queue<int, std::vector<int>> maxHeap;
  std::priority_queue<int, std::vector<int>, std::greater<int>> minHeap;

public:
  MedianFinder() {}

  void addNum(int num) {
    if (minHeap.empty() || (!maxHeap.empty() && num > maxHeap.top()) ||
        (!minHeap.empty() && num > minHeap.top()))
      minHeap.push(num);
    else
      maxHeap.push(num);

    if (minHeap.size() - maxHeap.size() == 2) {
      maxHeap.push(minHeap.top());
      minHeap.pop();
    } else if (maxHeap.size() - minHeap.size() == 1) {
      minHeap.push(maxHeap.top());
      maxHeap.pop();
    }
  }

  double findMedian() {
    if ((maxHeap.size() + minHeap.size()) % 2 == 0)
      return (double)(maxHeap.top() + minHeap.top()) / 2;
    else
      return (double)minHeap.top();
  }
};
